shared/src/main/scala/coop/rchain/store/LazyAdHocKeyValueCache.scala

```diff
@@ -5,42 +5,81 @@ import cats.effect.Concurrent
 import cats.effect.concurrent.{Deferred, Ref}
 import cats.syntax.all._
 
+import scala.collection.immutable.LinearSeq
+
 trait KeyValueCache[F[_], K, V] {
   def get(key: K, fallback: F[V]): F[V]
   def put(key: K, value: V): F[Unit]
   def toMap: F[Map[K, V]]
 }
 
+trait InvalidationPolicy[V] {
+  def processNewInsert(item: V): (Option[V], InvalidationPolicy[V])
+}
```

Comment:
Before we have clear understanding how limited cache will be implemented, defining some abstract definition will not make to much sense. Why this trait is necessary?

Reply:
To my understanding how cache is invalidated is something different to how cache is used. So I tried to write cache logic without implementation of invalidation. 

So invalidation check should happen after each new request for data that is not cached yet, resulting in expanding the cache. But how exactly this invalidation should influence cache - not known. 
Simplest is LRU cache with fixed map size, which is defined by order of item requested, and capacity of the seq (or set)
https://github.com/rchain/rchain/blob/bf08980d59c68f04fec14cee32518aa8aed9fc52/shared/src/main/scala/coop/rchain/store/LazyAdHocKeyValueCache.scala#L24-L33

What I found common for invalidation functions is that one should inputs some value (it might influence invalidation decision) and outputs option values to invalidate (can be set). Plus - the state that allows to make decision.

I tried first put everything in cache instance, but then trying to split task into different chunks, ended up with this solution.